specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/keys.json

```diff
@@ -0,0 +1,393 @@
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/Key"
+            }
+          },
+          "default": {
+            "schema": {
+              "$ref": "#/definitions/ErrorContract"
+            },
+            "description": "*** Error Responses: ***\n\n * 500 InternalServerError- Some error occurred while deleting Workspace Key."
+          }
+        },
+        "consumes": [],
+        "produces": [
+          "application/json"
+        ],
+        "x-ms-examples": {
+          "Patch a workspace key": {
+            "$ref": "./examples/UpdateKey.json"
+          }
+        }
+      }
+    }
+  },
+  "definitions": {
+    "ErrorDetail": {
+      "description": "Error details",
+      "type": "object",
+      "properties": {
+        "message": {
+          "description": "Error message",
+          "type": "string"
+        },
+        "code": {
+          "description": "Error code",
+          "type": "string"
+        },
+        "target": {
+          "description": "Error target",
+          "type": "string"
+        }
+      }
+    },
+    "ErrorContract": {
+      "title": "Error details.",
+      "description": "Contains details when the response code indicates an error.",
+      "type": "object",
+      "properties": {
+        "error": {
+          "description": "The error details.",
+          "$ref": "../../../../../common-types/resource-management/v1/types.json#/definitions/ErrorResponse"
+        }
+      }
+    },
+    "KeyInfoListResult": {
+      "description": "List of keys",
+      "type": "object",
+      "properties": {
+        "nextLink": {
+          "description": "Link to the next page of results",
+          "type": "string"
+        },
+        "value": {
+          "description": "List of keys",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/Key"
+          }
+        }
+      }
+    },
+    "Key": {
+      "description": "A workspace key",
+      "type": "object",
+      "allOf": [
+        {
+          "$ref": "../../../../../common-types/resource-management/v1/types.json#/definitions/TrackedResource"
+        }
+      ],
+      "properties": {
+        "properties": {
+          "description": "Keys resource properties",
+          "$ref": "#/definitions/KeyProperties",
+          "x-ms-client-flatten": true
+        }
+      }
+    },
+    "KeyPatchInfo": {
+      "description": "KeyInfo patch details",
+      "type": "object",
+      "properties": {
+        "tags": {
+          "description": "Resource tags",
+          "additionalProperties": {
+            "type": "string"
+          },
+          "type": "object"
+        }
+      }
+    },
+    "KeyProperties": {
+      "description": "Key properties",
+      "type": "object",
+      "properties": {
+        "isActiveCMK": {
+          "type": "boolean",
+          "description": "Used to activate the workspace after a customer managed key is provided."
```

Comment:
if these nested key resources are needed what does this boolean do?  The CMK setup is done on the workspace resource, not here

Reply:
We are doing a 2-step approach for CMK because we're going to do CMK via System-Assigned Managed Identity. Thus, we require the user to create the Synapse workspace resource and then 'activate' the resource through the Keys sub-resource after they've granted the System-Assigned Managed Identity with required permissions. We needed a signal through the sub-resource and this is what it's used for.

So here are the steps from a customer perspective:

1. Create the workspace resource (let's call it "someworkspace" with the encryption body: "cmk": {"key": {"name": "somekey", "kvurl": "https://vault.azure.net/keys/key1"}}
2. Add the SA-Managed Identity to your KeyVault's Access Policies
3. Do a PUT on the  created key-subresource (.../providers/Microsoft.Synapse/workspaces/someworkspace/keys/somekey)
{"isActiveCMK": true, "keyVaultUrl": "some valid kv url"}

We adopted this model because of 2 reasons:
- We're the evolution of SQL DW and they have keys subresources so we needed this subresource to future proof the merging of the 2 products
- Every time we do a PUT on the workspace itself, it is very expensive (reliability wise and performance wise) to do it through the workspace resource. Hence, the sub-resource allows us to have shorter flows through the sub-resource since no other changes are done in parallel.